### DISCUSSION TITLE
Mockup ideas for merging NewGRF/AI/GS Settings windows

### DISCUSSION THREAD

**Discussion by Andrew350** [OpenTTD/OpenTTD#8410] [General]
I'm not sure how much interest (if any) there might be for such a big change, but it makes sense to me that all "mods" should be joined under the same menu option, and I'm curious if others might agree or disagree. Let me know your thoughts. If nothing else the mockups might still work for a separate AI/GS window, if that's preferred.

Full disclaimer: I am not a programmer. Pretty pictures is all I can do, so this is just an example to potentially get some ideas going and maybe give some ambitious person with technical knowledge a starting point to actually undertake this.

The big challenge of merging the two windows aesthetically is that the NewGRF Settings window is more advanced and expandable, whereas the AI/GS window is much more confined and basic. So in a way this was more of an exercise in extending and reformatting the AI/GS window to the same standard so the two can coexist neatly and be more consistent with each other; adding buttons to switch between the two 'windows' is relatively simple (but of course that design choice is open for debate). How difficult this kind of change would be under-the-hood is beyond my level of knowledge, I just like the thought of having consistency between both menus.

I made mockups for both the default (small) and expanded windows, based on the current format of the NewGRF Settings window.

**Expanded:**
![mod_settings_mockup3](https://user-images.githubusercontent.com/50259667/102848133-e639cd00-43c8-11eb-8697-23125bd8c6e0.png)

**Minimum size:**
![mod_settings_mockup_small](https://user-images.githubusercontent.com/50259667/102848153-f5b91600-43c8-11eb-9dcc-11aeed555951.png)

The small window is a bit of a challenge to fit all the information in the same general format, so it looks a bit cluttered. This could perhaps be improved, but I think the only real solution to declutter it would be to use a different format altogether. Maybe something like the current implementation where a button must be pressed to show the AIs/GSs available, though that was one thing I specifically aimed to get rid of. Anyway it's all open to criticisms and suggestions to improve; this is just one idea.

Oh, and one other thing that would need to be addressed if these two menus get merged together is the now-empty button left on the main menu. I thought that would make a great place to put a link to the brand new OpenTTD Wiki, because why not?

![menu_mockup](https://user-images.githubusercontent.com/50259667/102848177-06698c00-43c9-11eb-8019-aa8d8a62989f.png)

**Comment by 2TallTyler** [comment]
I like this concept very much. The total menu separation has never made much sense to me, and it has resulted in me (and I suspect others) neglecting Game Scripts despite using NewGRFs -- out of sight, out of mind.

I also like the inclusion of the wiki link, although perhaps it should be titled Manual to tell players what they'll find there. The [manual of style](https://wiki.openttd.org/en/Wiki/Manual%20of%20Style#use-of-meta-words) on the wiki suggests this.

**Comment by James103** [comment]
This could also allow for a longer AI/GS description to fit on larger screens. Currently, the AI/GS description only fits if it's no more than 1-3 lines, but the upper limit could be increased to 16-20+ lines depending on the height of the window.

To take advantage of this while still allowing for a clear description on small window sizes, it is recommended to put a short description in the first couple lines, then a longer description (if necessary) in the next up to 10-16 or so lines.

**Comment by James103** [comment]
Just like NewGRFs have the active indicator show up blue when selected (but not active), green when active, nd red when disabled, a similar setup could be implemented for AI/GS:
* Blue: Pending
* Green: Active
* Yellow: Warning (NewGRF) or breakpoint (AI/GS) (optional)
* Red: Error/Fatal (NewGRF) or crashed script (AI/GS)

**Comment by michicc** [comment]
Somebody would have to dig up the threads on tt-forums, but many years ago there was discussion about reworking the new-game window. This is something I think would be a better target for unification.

The ideas back then were to have a multi-step window, e.g.:
1. Select basic settings like climate, map size etc (most of the existing new game window)
2. Select and configure AIs/GS
3. Select and configure NewGRFs
4. Configure all further game settings.
5. Start game.

You'd have a "Start now" button in each step to skip the remainder and just use the settings from last  time, like now.

**Reply by Andrew350** [reply]
I think I found the thread you're referring to:
https://www.tt-forums.net/viewtopic.php?f=32&t=46622

There are some interesting ideas there, albeit ten years out of date. For example Thgergo's mockup at the start of page 3 looks very similar to what I had in mind for a main menu with icons (minus the climate buttons); It's cool to see similar ideas from the past!

Regarding the new game GUI, I agree it should be more comprehensive and linear similar to other games; things are much too spread out right now. I'll play around with some ideas and see what I come up with, but a redesign of the AI/GS window seems to be a necessary step regardless, so hopefully that's still useful here.

**Reply by andythenorth** [reply]
See also https://www.tt-forums.net/viewtopic.php?f=33&t=48324

I had a bunch of mockups in there.  With the benefit of 10 years more experience, I wouldn't say they're great 😉

**Reply by Andrew350** [reply]
@andythenorth You wouldn't still happen to have any views of those other windows you were working on, would you? (or know where they might have been posted?)

**Reply by frosch123** [reply]
There are also
* https://wiki.openttd.org/en/Development/Design%20Drafts/GUI%20re-arrangement#new-game-dialog-tabs
* #5147
The latter also had some html mock-ups for the mapgen window.

Personally I think merging settings into the newgame window as tabs is a bad idea:
* The settings are also available in other places, so they still need separate windows.
* The mapgen window is in desperate need of other GUI items, like a height/mountain/sea/slope preview.
* Tabs assume that their content has the same window sizing behavior, which I don't think is the case.

Possibly even more off-topic, but I would like moving some buttons from the intro screen into the mapgen window.
* GS/AI/NewGRF settings make no sense in the intro screen.
* They do not apply to loading games, joining servers etc.
* They only apply to new games or starting servers (the latter also uses the newgame window).
* The scenario editor button is some weird mixture of loading scenarios or creating new ones, so it somewhat breaks the schema.

**Comment by SamuXarick** [comment]
I'd like to show my attempt at a reworked AI configuration window, where the main focus was to increase flexibility, removing some limitations that are currently imposed. So in a sence, I go against the idea of a unified GUI, which is what's being discussed here, because I feel that a AI is different than a NewGRF, it can be stopped, it can be changed. It's not set in stone. They are very different than each other to justify only 1 gui that suits both.

![687474703a2f2f692e696d6775722e636f6d2f4a6470466755352e706e67](https://user-images.githubusercontent.com/43006711/51566575-c8862500-1e8c-11e9-9486-514ff0c1ee4c.png)

A PR was issued, but ultimately rejected here: https://github.com/OpenTTD/OpenTTD/pull/7084

**Comment by frosch123** [comment]
I agree that GS and NewGRF are both mods that affect the game, and players do not need to know about their technical differences.

However, merging AI settings into mod settings makes no sense to me. While you have to download the them, they do not "mod" the gameplay.

**Reply by Andrew350** [reply]
@frosch123 Do you agree with michicc's proposal of unifying the New Game window?

If so, taking into account your opinion above, would you suggest that NewGRF and GS settings be grouped together while AI settings are left on a separate 'window', or rather they should all be separated into different steps, or some other combination?

**Reply by frosch123** [reply]
Uhm, I don't want to derail the discussion with too much off-topic stuff. But there are also ongoing discussions about
* Supporting more than one game-script.
* Bundling game-scripts and NewGRF into the same mod/file.

Neither of them is anywhere close to being done or even started, but if they were a thing:
* NewGRF and GS would probably share a "mod" list.
* NewGRF and GS are indistinguishable for a player, except that their errors looks different.

AIs would still be very different.

**Reply by EratoNysiad** [reply]
The NewGRFs and GS's are very different function wise however, even if we were to give them the same name I think they'd still be very much distinguishable because of that. It's nice to keep them somewhat segregated in that regard. I think having it like OP suggests, where you still separate NewGRFs and GS's, but still have them under the same button, is the best. They're both "mods", so it makes sense to put them together. But they're not the same, so it doesn't make sense to merge them completely.

Merging them the way you suggest sounds like something that will make old NewGRF downloads from the forums (for instance) no longer work, without providing any clear benefit to the player. Even if it didn't, it'd be very confusing for new NewGRF/GS developers who don't know the difference and thus don't know if they need NML or squirrel, or what the difference even is. With the clear distinction we have right now it's already hard enough for NewGRF developers not to quit before their first compile, we really can't use extra barriers to entry.

I agree with maybe separating AI and GS. ~~AI doesn't work to improve anyone's gameplay anyway.~~ It never made much sense to me that AI and GS were in the same window anyway.

**Reply by Andrew350** [reply]
I tend to agree with @frosch123 that NewGRF and GS share a similar role in changing gameplay (from a player perspective, anyway), so it seems sensible to "pair" GS with NewGRFs instead of AIs. Having yet another separate window for GS is a step backwards IMO, so if anything NewGRF and GS should share a configuration window and leave AIs as their own thing.

Here's a mockup example of how it could look:


![newgrf-gs-minimum](https://user-images.githubusercontent.com/50259667/103469111-70612a00-4d15-11eb-8a99-18bb73b2542a.png)


Basically just cut the GS part out of the AI settings window and add it to NewGRF window. This seems like something more easily achievable (and maybe more agreeable?) than an entire GUI redesign like unifying the new game window, while still moving in that general direction. IMO it's clean, simple, intuitive, scales nicely and doesn't seem to interfere with any potential future plans. If anything it may be a step towards other features like #8460, or unification of other windows, without hindering anything in the mean time. 

It would also make GSs more visible to players that might otherwise ignore them, so that's a positive 🙂 Of course AIs would then be left all alone, but an AI window makeover (like in my first mockup) becomes a bit more palatable without GS cluttering it up, so that could turn into a positive also.

**Reply by ConstanceHorne** [reply]
I see one more issue. I loaded file and it conflicted with other GRFs that I loaded and crashed my game so I removed it but now there is no way to actually _delete_ it from within the options window. It's just inactivated. I will never use it again. Where is the delete option? That seems like an omission. There should be an option to actually _delete_ GRFs that are not wanted. I like the two window idea. Makes sense to me.

**Reply by Andrew350** [reply]
> Where is the delete option?

I don't think deleting NewGRFs is a frequent enough thing to warrant adding a new button. You can always go to your newgrf folder and delete them manually if you feel the need to.

**Comment by ConstanceHorne** [comment]
Where exactly would I find them?

I checked under "documents > OpenTTD > newgrf" and there is nothing there.

Under "documents > OpenTTD > content_download > newgrf" I do see some tar
files. Is that what I am supposed to delete, a tar file?

If yes, then what is the other folder for?   "*documents > OpenTTD > newgrf*
"

**Reply by Andrew350** [reply]
> Under "documents > OpenTTD > content_download > newgrf" I do see some tar
> files. Is that what I am supposed to delete, a tar file?

Yep. Just make sure it's the correct one 😉 

> If yes, then what is the other folder for?   "*documents > OpenTTD > newgrf*

That's where you can place manually downloaded NewGRFs; content_download should only be used by the in-game content service.
